### DISCUSSION TITLE
Fybrik backward compatibility

### DISCUSSION THREAD

**Discussion by roytman** [fybrik/fybrik#1652] [Developers]
We as Fybrik developers should find a way to extend Fybrik functionality on the one hand and do not break its backward compatibility on the other hand. I'm opening this discussion to summarize all aspects of Fybrik's backward compatibility.
Below,  I described Kubernetes multi versions (CRDs) support. Everyone are welcome to add other backward compatability aspects.
 
### K8s API, Custom Resource Definitions (CRDs)
When we talk about backward compatibility, the first thing that is taking in account is CRDs. Kubernets provides mechanisms to support several apiVersions, and we are going to publish the first Fybrik  release with support two apiVersions( `v1alpha1` and `v1beta1`). 
There are two cases of multi versions supports: the versions with the same schema (like ours v1alpha1 and v1beta1), and when the versions have different schema. 
#### The same CRD schema
When the supported apiVersions have the same schema, multi-version support is quite straightforward. We have to:

- create an API  directory for a new version 
- define `conversion.strategy` as  `None`, - (The converter only change the apiVersion and would not touch any other field in the custom resource)
- define which version will be stored in etcd, can be only one
- if we use webhooks, we have to add them for a new version.   
- if we want, we can update the controllers to use a new (the stored) version.  The same for tests and examples.

  #### Support apiVersions with different schemas
When there are changes in the CRDs schemas, the things are more complicated. 

- In Kubernetes, all versions must be safely round-tripable through each other. This means that if we convert from version 1 to version 2, and then back to version 1, we must not lose information. Thus, any change we make to our API must be compatible with whatever we supported in v1, and also need to make sure anything we add in v2 is supported in v1.
- The most problematic issue is that CRDs schema conversation is done by Webhooks. I validated with the Kubernetes machinery team, there are no other options, except **Webhooks**. 
- We have to check how multiple CRD versions with different schemas work with taxonomy. Probably we have to include versions in the taxonomy validation. 
- Support of the conversations require additional steps, but they are straightforward.

**Comment by simanadler** [comment]
> We as Fybrik developers should find a way to extend Fybrik functionality on the one hand and do not break its backward compatibility on the other hand. I'm opening this discussion to summarize all aspects of Fybrik's backward compatibility. Below, I described Kubernetes multi versions (CRDs) support. Everyone are welcome to add other backward compatability aspects.
> ### K8s API, Custom Resource Definitions (CRDs)
> 
> When we talk about backward compatibility, the first thing that is taking in account is CRDs. Kubernets provides mechanisms to support several apiVersions, and we are going to publish the first Fybrik release with support two apiVersions( `v1alpha1` and `v1beta1`). There are two cases of multi versions supports: the versions with the same schema (like ours v1alpha1 and v1beta1), and when the versions have different schema.

[Sima] 
(1) When we released Fybrik 1.0 we explicitly stated that it will not support any previous versions, and that backward compatibility will only be supported moving forward.
(2) Why are we calling it beta1 instead of v1?
(3) What is the reason for having two different versions if nothing has changed?  Perhaps if we are (for example) trying to keep the FybrikApplication, plotter and blueprints aligned to the same version it makes sense.  Not clear whether we should or shouldn't be doing this.  What are the accepted approaches?

> #### The same CRD schema
> 
> When the supported apiVersions have the same schema, multi-version support is quite straightforward. We have to:
[Sima] Doing it may be straightforward, but when/why should we do so?
> 
>     * create an API  directory for a new version
> 
>     * define `conversion.strategy` as none `None`, - (The converter only change the apiVersion and would not touch any other field in the custom resource)
> 
>     * define which version will be stored in etcd, can be only one
> 
>     * if we use webhooks, we have to add them for a new version.
> 
>     * if we want, we can update the controllers to use a new (the stored) version.  The same for tests and examples.
>       #### Support apiVersions with different schemas
> 
> 
> When there are changes in the CRDs schemas, the things are more complicated.
> 
>     * In Kubernetes, all versions must be safely round-tripable through each other. This means that if we convert from version 1 to version 2, and then back to version 1, we must not lose information. Thus, any change we make to our API must be compatible with whatever we supported in v1, and also need to make sure anything we add in v2 is supported in v1.

[Sima] How is this even possible?  If we added new functionality in v2, it by definition didn't exist in v1.
> 
>     * The most problematic issue is that CRDs schema conversation is done by Webhooks. I validated with the Kubernetes machinery team, there are no other options, except **Webhooks**.

[Sima] If this is the case we need to discuss with our current and potential customers.  No point in declaring  support for versioning if it can only be done via webhooks and they can't use webhooks

> 
>     * We have to check how multiple CRD versions with different schemas work with taxonomy. Probably we have to include versions in the taxonomy validation.

[Sima] Taxonomy, connectors, CSP i/f, modules, IT config policies, ...  There are lots of things that aren't CRDs and we need a comprehensive approach about both when and how to update versions and how to support older ones.
> 
>     * Support of the conversations require additional steps, but they are straightforward.

**Comment by roytman** [comment]
> Sima] How is this even possible? If we added new functionality in v2, it by definition didn't exist in v1.
You can store the extended info in V1 as for example annotations.

> [Sima] If this is the case we need to discuss with our current and potential customers. No point in declaring support for versioning if it can only be done via webhooks and they can't use webhooks
We have to think twice before changing CRDs schemas. Not all changes actually change the schema. And yes, we have to talk with the customers.

> [Sima] Taxonomy, connectors, CSP i/f, modules, IT config policies, ... There are lots of things that aren't CRDs and we need a comprehensive approach about both when and how to update versions and how to support older ones.

I have started this discussion to aggregate all aspects of compatibility,  and explicitly mentioned  that I'm starting with CRDs, but  there are other aspects.
